ignite/metrics/rouge.py

```diff
@@ -0,0 +1,178 @@
+import numbers
+from collections import defaultdict
+from typing import Callable, DefaultDict, List, Tuple, Union
+
+import torch
+from torch.types import Number
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics import Metric
+
+# These decorators helps with distributed settings
+from ignite.metrics.metric import reinit__is_reduced, sync_all_reduce
+
+
+def _ngramify(tokens: List[str], n: int) -> DefaultDict:
+    ngrams: DefaultDict = defaultdict(int)
+    for ngram in (tuple(tokens[i : i + n]) for i in range(len(tokens) - n + 1)):
+        ngrams[ngram] += 1
+    return ngrams
+
+
+def _safe_divide(numerator: float, denominator: float) -> float:
+    if denominator > 0:
+        return numerator / denominator
+    else:
+        return 0.0
+
+
+def _fbeta_score(matches: int, recall_total: int, precision_total: int, beta_sq: float) -> float:
+    recall_score = _safe_divide(matches, recall_total)
+    precision_score = _safe_divide(matches, precision_total)
+    denom = (beta_sq) * precision_score + recall_score
+    fbeta_score = _safe_divide((1 + beta_sq) * precision_score * recall_score, denom)
+    return fbeta_score
+
+
+def _lcs(a: List[str], b: List[str]) -> int:
+    if len(a) < len(b):
+        a, b = b, a
+
+    if len(b) == 0:
+        return 0
+
+    row = [0] * len(b)
+    for ai in a:
+        left = 0
+        diag = 0
+        for j, bj in enumerate(b):
+            up = row[j]
+            if ai == bj:
+                value = diag + 1
+            else:
+                value = max(left, up)
+            row[j] = value
+            left = value
+            diag = up
+    return left
+
+
+class Rouge(Metric):
+    r"""Calculates the Rouge Score for two Sequence of Tokens.
+
+    Paper: Lin, Chin-Yew. 2004. ROUGE: a Package for Automatic Evaluation of Summaries.
+    In Proceedings of the Workshop on Text Summarization Branches Out (WAS 2004), Barcelona, Spain, July 25 - 26, 2004.
+
+    Contains two metric Rouge-n and Rouge-L.
+
+    - Rouge-n: precision, recall and f-beta based on overlapping n-grams between the predicted and the reference text
+    - Rouge-L: Only the sentence level Rouge-L score is implemented. It finds the Longest Common Subsequence between
+      the predicted and reference text and uses it to calculate precison, recall, f-beta score
+
+    .. math::
+        F_\beta = \left( 1 + \beta^2 \right) * \frac{ \text{precision} * \text{recall} }
+        { \left( \beta^2 * \text{precision} \right) + \text{recall} }
+
+    where :math:`\beta` is a float.
+
+    - ``update`` must receive output of the form ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y}``.
+    - `y_pred` must be a Sequence of Tokens.
+    - `y` must be in the form of List of Sequence of Tokens of Model Texts.
+
+    Args:
+        beta: beta value for calculating f-beta
+        variant: Which variant of Rouge Score to be evaluated
+            Valid Values - "rougeN", "rougen", "n", "N", "rougeL", "rougel", "l", "L"
```

Comment:
This is not understandable... Too many ways to define the same algorithm.

For instance (think about the full version) 
```suggestion
            Valid Values - "corpus-N", "sentence-N", "corpus-L", "sentence-L", 
```
Some arguments should be mutual exclusive.

Reply:
Please have a look of the notebook I send yesterday.